File: ios/MagicFormsMobile/Services/AuthAPI.swift
import Foundation

extension Notification.Name {
    static let mobileSessionExpired = Notification.Name("MobileSessionExpired")
}

enum AuthAPI {
    private static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        return d
    }()

    private static let encoder: JSONEncoder = {
        let e = JSONEncoder()
        return e
    }()

    static func getData(url: URL, token: String?) async throws -> Data {
        try await get(url: url, token: token)
    }

    static func postData(url: URL, body: Data, token: String?) async throws -> Data {
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        applyCommonHeaders(&request, token: token)
        request.httpBody = body
        return try await perform(request)
    }

    static func deleteData(url: URL, token: String?) async throws -> Data {
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        applyCommonHeaders(&request, token: token)
        return try await perform(request)
    }

    static func postMultipartForm(
        url: URL,
        token: String?,
        textFields: [String: String],
        checkboxFields: [String: Bool] = [:],
        checklistFields: [String: [String]] = [:],
        files: [String: (data: Data, fileName: String, mimeType: String)] = [:]
    ) async throws -> Data {
        let boundary = "Boundary-\(UUID().uuidString)"
        var body = Data()

        func appendField(name: String, value: String) {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n".data(using: .utf8)!)
            body.append("\(value)\r\n".data(using: .utf8)!)
        }

        for (key, value) in textFields {
            appendField(name: key, value: value)
        }
        for (key, checked) in checkboxFields where checked {
            appendField(name: key, value: "on")
        }
        for (key, values) in checklistFields {
            for value in values {
                appendField(name: key, value: value)
            }
        }
        for (fieldName, file) in files {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append(
                "Content-Disposition: form-data; name=\"\(fieldName)\"; filename=\"\(file.fileName)\"\r\n"
                    .data(using: .utf8)!
            )
            body.append("Content-Type: \(file.mimeType)\r\n\r\n".data(using: .utf8)!)
            body.append(file.data)
            body.append("\r\n".data(using: .utf8)!)
        }
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        applyCommonHeaders(&request, token: token)
        request.httpBody = body
        return try await perform(request)
    }

    static func uploadMultipart(
        url: URL,
        token: String?,
        fields: [String: String],
        fileField: String,
        fileData: Data,
        fileName: String,
        mimeType: String
    ) async throws -> Data {
        let boundary = "Boundary-\(UUID().uuidString)"
        var body = Data()
        for (key, value) in fields {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"\(key)\"\r\n\r\n".data(using: .utf8)!)
            body.append("\(value)\r\n".data(using: .utf8)!)
        }
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append(
            "Content-Disposition: form-data; name=\"\(fileField)\"; filename=\"\(fileName)\"\r\n"
                .data(using: .utf8)!
        )
        body.append("Content-Type: \(mimeType)\r\n\r\n".data(using: .utf8)!)
        body.append(fileData)
        body.append("\r\n".data(using: .utf8)!)
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        applyCommonHeaders(&request, token: token)
        request.httpBody = body
        return try await perform(request)
    }

    static func fetchDirectoryEntities() async throws -> [APIEntity] {
        let url = APIConfig.url(path: "/api/v1/auth/directory-entities/")
        let data = try await get(url: url, token: nil)
        let response = try decoder.decode(DirectoryEntitiesResponse.self, from: data)
        return response.entities ?? []
    }

    static func login(username: String, password: String, entitySlug: String?) async throws -> LoginResponse {
        let url = APIConfig.url(path: "/api/v1/auth/login/")
        let body = LoginRequest(
            username: username,
            password: password,
            entitySlug: entitySlug?.isEmpty == true ? nil : entitySlug
        )
        let data = try await post(url: url, body: body, token: nil)
        let response = try decoder.decode(LoginResponse.self, from: data)
        if response.ok, response.token != nil {
            return response
        }
        throw try parseError(data: data)
    }

    static func me(token: String) async throws -> MeResponse {
        let url = APIConfig.url(path: "/api/v1/auth/me/")
        let data = try await get(url: url, token: token)
        let response = try decoder.decode(MeResponse.self, from: data)
        if response.ok { return response }
        throw try parseError(data: data)
    }

    static func logout(token: String) async throws {
        let url = APIConfig.url(path: "/api/v1/auth/logout/")
        _ = try await post(url: url, body: EmptyBody(), token: token)
    }

    // MARK: - HTTP

    private struct EmptyBody: Encodable {}

    private static func get(url: URL, token: String?) async throws -> Data {
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        applyCommonHeaders(&request, token: token)
        return try await perform(request)
    }

    private static func post<T: Encodable>(url: URL, body: T, token: String?) async throws -> Data {
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        applyCommonHeaders(&request, token: token)
        request.httpBody = try encoder.encode(body)
        return try await perform(request)
    }

    private static func applyCommonHeaders(_ request: inout URLRequest, token: String?) {
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(AppLanguage.currentAcceptLanguage, forHTTPHeaderField: "Accept-Language")
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
    }

    private static func perform(_ request: URLRequest) async throws -> Data {
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw APIError(code: "network", message: "Invalid response.", entities: nil)
        }
        if (200 ..< 300).contains(http.statusCode) {
            return data
        }
        if http.statusCode == 401 {
            NotificationCenter.default.post(name: .mobileSessionExpired, object: nil)
        }
        throw try parseError(data: data, status: http.statusCode)
    }

    static func decodeError(from data: Data, status: Int? = nil) throws -> APIError {
        try parseError(data: data, status: status)
    }

    private static func parseError(data: Data, status: Int? = nil) throws -> APIError {
        if let body = try? decoder.decode(APIErrorBody.self, from: data) {
            let code = body.error ?? "error"
            let message = body.message ?? "Request failed."
            return APIError(code: code, message: message, entities: body.entities)
        }
        if let status {
            return APIError(code: "http_\(status)", message: "HTTP \(status)", entities: nil)
        }
        return APIError(code: "error", message: "Request failed.", entities: nil)
    }
}

File: android/app/src/main/java/com/magicforms/mobile/data/api/AuthApi.kt
package com.magicforms.mobile.data.api

import com.magicforms.mobile.data.models.ApiEntity
import com.magicforms.mobile.data.models.ApiErrorBody
import com.magicforms.mobile.data.models.ApiException
import com.magicforms.mobile.data.models.DirectoryEntitiesResponse
import com.magicforms.mobile.data.models.LoginRequest
import com.magicforms.mobile.data.models.LoginResponse
import com.magicforms.mobile.data.LanguagePreferences
import com.magicforms.mobile.data.SessionExpiredException
import com.magicforms.mobile.data.models.MeResponse
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.MultipartBody
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

class AuthApi {
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    private val client = OkHttpClient.Builder().build()
    private val jsonMedia = "application/json; charset=utf-8".toMediaType()

    suspend fun fetchDirectoryEntities(): List<ApiEntity> {
        val data = get(ApiConfig.url("/api/v1/auth/directory-entities/"), token = null)
        val response = json.decodeFromString<DirectoryEntitiesResponse>(data)
        return response.entities.orEmpty()
    }

    suspend fun login(username: String, password: String, entitySlug: String?): LoginResponse {
        val body = LoginRequest(
            username = username,
            password = password,
            entitySlug = entitySlug?.takeIf { it.isNotBlank() },
        )
        val data = postRequest(ApiConfig.url("/api/v1/auth/login/"), json.encodeToString(body), token = null)
        val response = json.decodeFromString<LoginResponse>(data)
        if (response.ok && !response.token.isNullOrBlank()) {
            return response
        }
        throw parseError(data)
    }

    suspend fun me(token: String): MeResponse {
        val data = get(ApiConfig.url("/api/v1/auth/me/"), token = token)
        val response = json.decodeFromString<MeResponse>(data)
        if (response.ok) return response
        throw parseError(data)
    }

    suspend fun logout(token: String) {
        postRequest(ApiConfig.url("/api/v1/auth/logout/"), "{}", token = token)
    }

    fun get(url: String, token: String?): String = getRequest(url, token)

    fun post(url: String, body: String, token: String?): String = postRequest(url, body, token)

    fun postMultipart(url: String, body: MultipartBody, token: String?): String {
        val request = Request.Builder()
            .url(url)
            .post(body)
            .header("Accept", "application/json")
            .apply { token?.let { header("Authorization", "Bearer $it") } }
            .build()
        return execute(request)
    }

    fun delete(url: String, token: String?): String {
        val request = applyLanguage(
            Request.Builder()
                .url(url)
                .delete()
                .header("Accept", "application/json")
                .apply { token?.let { header("Authorization", "Bearer $it") } },
        ).build()
        return execute(request)
    }

    private fun getRequest(url: String, token: String?): String {
        val request = applyLanguage(
            Request.Builder()
                .url(url)
                .get()
                .header("Accept", "application/json")
                .apply { token?.let { header("Authorization", "Bearer $it") } },
        ).build()
        return execute(request)
    }

    private fun postRequest(url: String, body: String, token: String?): String {
        val request = applyLanguage(
            Request.Builder()
                .url(url)
                .post(body.toRequestBody(jsonMedia))
                .header("Content-Type", "application/json")
                .header("Accept", "application/json")
                .apply { token?.let { header("Authorization", "Bearer $it") } },
        ).build()
        return execute(request)
    }

    private fun applyLanguage(request: Request.Builder): Request.Builder {
        return request.header("Accept-Language", LanguagePreferences.acceptLanguageHeader())
    }

    private fun execute(request: Request): String {
        client.newCall(request).execute().use { response ->
            val body = response.body?.string().orEmpty()
            if (response.isSuccessful) return body
            if (response.code == 401) {
                throw SessionExpiredException()
            }
            throw parseError(body, response.code)
        }
    }

    private fun parseError(data: String, status: Int? = null): ApiException {
        return try {
            val body = json.decodeFromString<ApiErrorBody>(data)
            ApiException(
                code = body.error ?: "error",
                message = body.message ?: "Request failed.",
                entities = body.entities,
            )
        } catch (_: Exception) {
            ApiException(
                code = status?.let { "http_$it" } ?: "error",
                message = status?.let { "HTTP $it" } ?: "Request failed.",
            )
        }
    }
}
